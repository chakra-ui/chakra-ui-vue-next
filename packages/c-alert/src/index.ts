import { defineComponent, computed, createVNode, mergeProps, isVNode, Fragment } from 'vue';
import { createStylesContext, useMultiStyleConfig, chakra } from '@chakra-ui/vue-system';
import { createContext, getValidChildren } from '@chakra-ui/vue-utils';
import { createIconComponent } from '@chakra-ui/c-icon';

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };
  return _extends.apply(this, arguments);
}

var CInfoIcon = createIconComponent("info");
var CCheckIcon = createIconComponent("check-circle");
var CWarningIcon = createIconComponent("warning-alt");
var CErrorIcon = createIconComponent("warning");

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}

var STATUSES = {
  info: {
    colorScheme: "blue",
    icon: CInfoIcon
  },
  success: {
    colorScheme: "green",
    icon: CCheckIcon
  },
  warning: {
    colorScheme: "orange",
    icon: CWarningIcon
  },
  error: {
    colorScheme: "red",
    icon: CErrorIcon
  },
  loading: {
    icon: CInfoIcon,
    colorScheme: "blue"
  }
};

var _createStylesContext = createStylesContext("Alert"),
    StylesProvider = _createStylesContext[0],
    useStyles = _createStylesContext[1];

var _createContext = createContext({
  name: "AlertContext",
  errorMessage: "useAlertContext: `context` is undefined. Seems you forgot to wrap alert components in `<c-alert />`"
}),
    AlertProvider = _createContext[0],
    useAlertContext = _createContext[1];
/**
 * CAlert component
 *
 * This is the container component for all Alert components.
 * It also provides state and context to it's compound components
 */


var CAlert = defineComponent({
  name: "CAlert",
  props: {
    as: {
      type: [String, Object],
      "default": "div"
    },
    status: {
      type: [String],
      "default": "info"
    },
    colorScheme: {
      type: [String]
    },
    styleConfig: {
      type: [Object]
    },
    variant: {
      type: [String],
      "default": "solid"
    }
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var colorScheme = computed(function () {
      var _STATUSES$props$statu;

      return props.colorScheme || ((_STATUSES$props$statu = STATUSES[props == null ? void 0 : props.status]) == null ? void 0 : _STATUSES$props$statu.colorScheme);
    });
    var themingProps = computed(function () {
      return {
        colorScheme: colorScheme.value,
        variant: props.variant
      };
    });
    AlertProvider({
      status: computed(function () {
        return props.status;
      })
    });
    var styles = useMultiStyleConfig("Alert", themingProps);
    StylesProvider(styles);
    var alertStyles = computed(function () {
      return _extends({
        width: "100%",
        display: "flex",
        alignItems: "center",
        position: "relative",
        overflow: "hidden"
      }, styles.value.container);
    });
    return function () {
      return createVNode(chakra.div, mergeProps({
        "role": "alert",
        "__label": "alert",
        "__css": alertStyles.value
      }, attrs), {
        "default": function _default() {
          return getValidChildren(slots);
        }
      });
    };
  }
});
/**
 * CAlertTitle component
 *
 * The title component for alerts
 */

var CAlertTitle = defineComponent({
  name: "CAlertTitle",
  setup: function setup(_, _ref2) {
    var attrs = _ref2.attrs,
        slots = _ref2.slots;
    var styles = useStyles();
    return function () {
      return createVNode(chakra.div, mergeProps({
        "__label": "alert__title",
        "__css": styles.value.title
      }, attrs), _isSlot(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});
/**Â«
 * CAlertDescription component
 *
 * The description component for alerts
 */

var CAlertDescription = defineComponent({
  name: "CAlertDescription",
  setup: function setup(_, _ref3) {
    var attrs = _ref3.attrs,
        slots = _ref3.slots;
    var styles = useStyles();
    return function () {
      return createVNode(chakra.div, mergeProps({
        "__label": "alert__description",
        "__css": styles.value.description
      }, attrs), {
        "default": function _default() {
          return getValidChildren(slots);
        }
      });
    };
  }
});
/**
 * CAlertIcon component
 *
 * The Icon component for alerts
 */

var CAlertIcon = defineComponent({
  name: "CAlertIcon",
  setup: function setup(_, _ref4) {
    var attrs = _ref4.attrs,
        slots = _ref4.slots;
    var styles = useStyles();

    var _useAlertContext = useAlertContext(),
        status = _useAlertContext.status;

    var BaseIcon = STATUSES[status.value].icon;
    var css = computed(function () {
      return status.value === "loading" ? styles.value.spinner : styles.value.icon;
    });
    return function () {
      var validChildren = getValidChildren(slots);
      return createVNode(chakra.span, mergeProps({
        "display": "inherit",
        "__label": "alert__icon"
      }, attrs, {
        "__css": css.value
      }), {
        "default": function _default() {
          return createVNode(Fragment, null, [validChildren.length ? slots : createVNode(BaseIcon, {
            "h": "100%",
            "w": "100%"
          }, null)]);
        }
      }); // return <icon {...styles.value.icon} {...attrs}></icon>
    };
  }
});

export { CAlert, CAlertDescription, CAlertIcon, CAlertTitle };
