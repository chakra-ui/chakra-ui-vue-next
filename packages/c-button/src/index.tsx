import { defineComponent, computed, createVNode, mergeProps, Fragment, createTextVNode, isVNode } from 'vue';
import { chakra, useStyleConfig } from '@chakra-ui/vue-system';
import { mergeWith, filterUndefined, dataAttr } from '@chakra-ui/utils';
import { vueThemingProps, createContext, getValidChildren, SNAO } from '@chakra-ui/vue-utils';
import { CIcon } from '@chakra-ui/c-icon';
import { CSpinner } from '@chakra-ui/c-spinner';

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };
  return _extends.apply(this, arguments);
}

var props = _extends({
  isAttached: Boolean,
  isDisabled: Boolean,
  spacing: {
    type: [String, Number, Array],
    "default": 3
  }
}, vueThemingProps);

var _createContext = createContext({
  strict: false,
  name: "ButtonGroupContext"
}),
    ButtonGroupProvider = _createContext[0],
    useButtonGroup = _createContext[1];

var CButtonGroup = defineComponent({
  name: "CButtonGroup",
  props: props,
  setup: function setup(props, _ref) {
    var attrs = _ref.attrs,
        slots = _ref.slots;
    ButtonGroupProvider( // @ts-ignore
    computed(function () {
      return {
        size: props.size,
        colorScheme: props.colorScheme,
        variant: props.variant,
        isDisabled: props.isDisabled
      };
    }));
    var styles = computed(function () {
      var groupStyles = {
        display: "inline-flex"
      };

      if (props.isAttached) {
        groupStyles = _extends({}, groupStyles, {
          "> *:first-of-type:not(:last-of-type)": {
            borderRightRadius: 0
          },
          "> *:not(:first-of-type):not(:last-of-type)": {
            borderRadius: 0
          },
          "> *:not(:first-of-type):last-of-type": {
            borderLeftRadius: 0
          }
        });
      } else {
        groupStyles = _extends({}, groupStyles, {
          "& > *:not(style) ~ *:not(style)": {
            marginLeft: props.spacing
          }
        });
      }

      return groupStyles;
    });
    return function () {
      return createVNode(chakra.div, mergeProps({
        "role": "group",
        "__label": "button__group",
        "__css": styles.value
      }, attrs), {
        "default": function _default() {
          return getValidChildren(slots);
        }
      });
    };
  }
});

var defaultButtonProps = {
  as: "button",
  iconSpacing: "0.5rem"
};

var CButtonSpinner = defineComponent({
  name: "CButtonSpinner",
  props: {
    label: Boolean,
    spacing: [Number, String, Array],
    placement: String
  },
  setup: function setup(props, _ref) {
    var attrs = _ref.attrs;
    var marginProp = computed(function () {
      return props.placement === "start" ? "marginEnd" : "marginStart";
    });
    var spinnerStyles = computed(function () {
      var _ref2;

      return _ref2 = {
        display: "flex",
        alignItems: "center",
        position: props.label ? "relative" : "absolute"
      }, _ref2[marginProp.value] = props.label ? props.spacing || "0.5rem" : "0", _ref2;
    });
    return function () {
      return (// @ts-ignore
        createVNode(chakra.div, mergeProps({
          "__label": "button__spinner"
        }, spinnerStyles.value, attrs), {
          "default": function _default() {
            return [createVNode(CSpinner, {
              "width": "1em",
              "height": "1em"
            }, null)];
          }
        })
      );
    };
  }
});
var CButtonContent = defineComponent({
  name: "CButtonContent",
  props: {
    leftIcon: String,
    rightIcon: String,
    iconSpacing: String
  },
  setup: function setup(props, _ref3) {
    var slots = _ref3.slots;
    return function () {
      return createVNode(Fragment, null, [props.leftIcon && createVNode(CButtonIcon, {
        "icon": props.leftIcon,
        "marginEnd": props.iconSpacing
      }, null), slots == null ? void 0 : slots["default"] == null ? void 0 : slots["default"](), props.rightIcon && createVNode(CButtonIcon, {
        "icon": props.rightIcon,
        "marginStart": props.iconSpacing
      }, null)]);
    };
  }
});
/**
 * CButtonIcon
 *
 * Button icon component
 */

var CButtonIcon = defineComponent({
  name: "CButtonIcon",
  props: {
    icon: String
  },
  setup: function setup(props, _ref4) {
    var attrs = _ref4.attrs;
    return function () {
      return createVNode(CIcon, mergeProps({
        "__label": "button__icon",
        "name": props.icon
      }, attrs), null);
    };
  }
});

/**
 * CButton
 *
 * The Button component is used to trigger an action or event,
 * such as submitting a form, opening a dialog, canceling
 * an action, or performing a delete operation.
 */
var CButton = defineComponent({
  name: "CButton",
  props: _extends({
    isLoading: {
      type: Boolean
    },
    isDisabled: {
      type: Boolean
    },
    isActive: {
      type: Boolean
    },
    loadingText: {
      type: String
    },
    isFullWidth: {
      type: Boolean
    },
    type: {
      type: String
    },
    leftIcon: {
      type: String
    },
    rightIcon: {
      type: String
    },
    iconSpacing: {
      type: SNAO
    },
    spinnerPlacement: {
      type: String,
      "default": "start"
    }
  }, vueThemingProps),
  setup: function setup(_props, _ref5) {
    var attrs = _ref5.attrs,
        slots = _ref5.slots;
    var props = computed(function () {
      return mergeWith({}, defaultButtonProps, _props, attrs);
    });
    var themingProps = computed(function () {
      return filterUndefined({
        colorScheme: props.value.colorScheme,
        variant: props.value.variant,
        size: props.value.size,
        styleConfig: props.value.styleConfig
      });
    });
    var group = useButtonGroup();
    var styles = useStyleConfig("Button", computed(function () {
      return _extends({}, group == null ? void 0 : group.value, themingProps.value);
    }));

    var _focus = computed(function () {
      var _styles$value$_focus, _styles$value;

      return mergeWith({}, (_styles$value$_focus = (_styles$value = styles.value) == null ? void 0 : _styles$value["_focus"]) != null ? _styles$value$_focus : {}, {
        zIndex: 1
      });
    });

    var buttonStyles = computed(function () {
      return _extends({
        display: "inline-flex",
        appearance: "none",
        alignItems: "center",
        justifyContent: "center",
        transition: "all 250ms",
        userSelect: "none",
        position: "relative",
        whiteSpace: "nowrap",
        verticalAlign: "middle",
        outline: "none",
        width: props.value.isFullWidth ? "100%" : "auto"
      }, styles.value, !!(group != null && group.value) && {
        _focus: _focus.value
      });
    });
    return function () {
      return createVNode(chakra.button, mergeProps({
        "as": props.value.as,
        "label": "button"
      }, (props.value.isDisabled || props.value.isLoading) && {
        disabled: props.value.isDisabled || props.value.isLoading
      }, {
        "type": props.value.as === "button" ? undefined : props.value.type,
        "data-active": dataAttr(props.value.isActive),
        "data-loading": dataAttr(props.value.isLoading),
        "__css": buttonStyles.value
      }, attrs), {
        "default": function _default() {
          return createVNode(Fragment, null, [props.value.isLoading && props.value.spinnerPlacement === "start" && createVNode(CButtonSpinner, {
            "placement": "start",
            "spacing": props.value.iconSpacing,
            "__label": "button-spinner__start",
            "label": props.value.loadingText,
            "__css": {
              fontSize: "1em",
              lineHeight: "normal"
            }
          }, null), props.value.isLoading ? props.value.loadingText || createVNode(chakra.span, {
            "opacity": 0
          }, {
            "default": function _default() {
              return [createVNode(CButtonContent, {
                "leftIcon": props.value.leftIcon,
                "rightIcon": props.value.rightIcon,
                "iconSpacing": props.value.iconSpacing
              }, {
                "default": function _default() {
                  return [slots == null ? void 0 : slots["default"] == null ? void 0 : slots["default"]()];
                }
              })];
            }
          }) : createVNode(CButtonContent, {
            "leftIcon": props.value.leftIcon,
            "rightIcon": props.value.rightIcon,
            "iconSpacing": props.value.iconSpacing
          }, {
            "default": function _default() {
              return getValidChildren(slots);
            }
          }), props.value.isLoading && props.value.spinnerPlacement === "end" && createVNode(CButtonSpinner, {
            "placement": "end",
            "spacing": props.value.iconSpacing,
            "__label": "button-spinner__end",
            "label": props.value.loadingText,
            "__css": {
              fontSize: "1em",
              lineHeight: "normal"
            }
          }, null)]);
        }
      });
    };
  }
});

var IconButtonProps = {
  // ...BUTTON_PROPS,
  icon: String,
  isRound: Boolean,
  ariaLabel: {
    type: String,
    required: true
  }
};

/**
 * CIconButton
 *
 * IconButton composes the Button component except that it renders only an icon.
 */
var CIconButton = defineComponent({
  name: "CIconButton",
  props: IconButtonProps,
  setup: function setup(props, _ref) {
    var attrs = _ref.attrs;

    if (!props.ariaLabel) {
      console.error("chakra-ui: The `aria-label` prop is required for the <c-icon-button />");
    }

    return function () {
      return createVNode(CButton, mergeProps({
        "padding": "0",
        "rounded": props.isRound ? "rounded" : "md",
        "aria-label": props.ariaLabel
      }, attrs), {
        "default": function _default() {
          return [createVNode(CIcon, {
            "aria-hidden": true,
            "focusable": 0,
            "name": props.icon
          }, null)];
        }
      });
    };
  }
});
var Foo = defineComponent(function (props, _ref2) {
  var slots = _ref2.slots;
  return function () {
    return createVNode("div", null, [createVNode("span", {
      "data-foo": true
    }, [createTextVNode("Hello")]), slots]);
  };
});

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}
defineComponent(function (_, _ref) {
  var slots = _ref.slots;
  return function () {
    return createVNode(Foo, null, _isSlot(slots) ? slots : {
      "default": function _default() {
        return [slots];
      }
    });
  };
});

export { CButton, CButtonGroup, CIconButton, Foo };
