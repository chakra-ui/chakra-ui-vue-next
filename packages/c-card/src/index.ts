export { CCard, type CCardProps } from "./c-card"
export { CCardBody, type CCardBodyProps } from "./c-card-body"
export { CCardHeader, type CCardHeaderProps } from "./c-card-header"
