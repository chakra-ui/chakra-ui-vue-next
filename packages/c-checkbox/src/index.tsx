export { CCheckbox } from "./checkbox"
export {
  CCheckboxGroup,
  useCheckboxGroupContext,
  type CheckboxGroupContext,
} from "./checkbox-group"

export type { CCheckboxControlProps, CCheckboxProps } from "./checkbox"
export type { CCheckboxGroupProps } from "./checkbox-group"
