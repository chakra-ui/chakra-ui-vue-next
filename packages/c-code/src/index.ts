export { default as CCode } from './code'
