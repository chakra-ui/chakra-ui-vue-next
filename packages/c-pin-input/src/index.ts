import { computed, watch, defineComponent, createVNode, mergeProps, h } from 'vue';
import { createStylesContext, useStyleConfig, chakra } from '@chakra-ui/vue-system';
import { createContext, vueThemingProps, getValidChildren } from '@chakra-ui/vue-utils';
import { CInput } from '@chakra-ui/c-input';
import { machine, connect } from '@zag-js/pin-input';
import { useMachine, normalizeProps } from '@zag-js/vue';

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };
  return _extends.apply(this, arguments);
}

var usePinInputMachine = function usePinInputMachine(params, emit) {
  var _useMachine = useMachine(machine({
    id: params.id,
    value: params.modelValue || params.value,
    placeholder: params.placeholder,
    type: params.type,
    otp: params.otp,
    mask: params.mask,
    blurOnComplete: params.blurOnComplete,
    onChange: function onChange(value) {
      emit("change", value);
      emit("update:modelValue", value);
    },
    onComplete: function onComplete(_ref) {
      var value = _ref.value,
          valueAsString = _ref.valueAsString;
      emit("complete", {
        value: value,
        valueAsString: valueAsString
      });
    },
    onInvalid: function onInvalid(details) {
      emit("invalid", details);
    }
  })),
      state = _useMachine[0],
      send = _useMachine[1];

  var api = computed(function () {
    return connect(state.value, send, normalizeProps);
  });
  watch(function () {
    return params.modelValue;
  }, function (value) {
    api.value.setValue(Array.from(value));
  });
  return api;
};

var _createStylesContext = createStylesContext("CPinInput"),
    StylesProvider = _createStylesContext[0],
    useStyles = _createStylesContext[1];

var _createContext = createContext(),
    PinInputProvider = _createContext[0],
    usePinInput = _createContext[1];
var CPinInputProps = _extends({
  value: {
    type: Object,
    "default": []
  },
  modelValue: {
    type: Object,
    "default": false
  },
  id: {
    type: String,
    "default": "0"
  },
  placeholder: {
    type: String,
    "default": "o"
  },
  type: {
    type: String,
    "default": "numeric"
  },
  otp: {
    type: Boolean,
    "default": false
  },
  mask: {
    type: Boolean,
    "default": false
  },
  blurOnComplete: {
    type: Boolean,
    "default": false
  },
  dir: {
    type: String,
    "default": "ltr"
  },
  spacing: {
    type: [String, Number],
    "default": "0.75"
  }
}, vueThemingProps);
var CPinInput = defineComponent({
  name: "CPinInput",
  props: CPinInputProps,
  emits: ["change", "invalid", "complete", "update:modelValue"],
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs,
        emit = _ref.emit;
    var styles = useStyleConfig("PinInput", props);
    var inputStyles = computed(function () {
      return _extends({}, styles.value, {
        mx: props.spacing
      });
    });
    StylesProvider(inputStyles);
    var api = usePinInputMachine(props, emit);
    PinInputProvider(api);
    return function () {
      return createVNode(chakra.div, mergeProps({
        "__label": "pin-input"
      }, api.value.rootProps, attrs), {
        "default": function _default() {
          return getValidChildren(slots).map(function (child, index) {
            return child.type.name === "CPinInputField" ? h(child, {
              index: index
            }) : child;
          });
        }
      });
    };
  }
});
var CPinInputField = defineComponent({
  name: "CPinInputField",
  props: {
    index: {
      type: Number,
      "default": 0
    }
  },
  setup: function setup(props, _ref2) {
    var attrs = _ref2.attrs;
    var styles = useStyles();
    var api = usePinInput();
    return function () {
      return createVNode(CInput, mergeProps({
        "__label": "pin-input-field",
        "__css": styles.value
      }, api.value.getInputProps({
        index: props.index
      }), attrs), null);
    };
  }
});
var CPinInputClearButton = defineComponent({
  name: "CPinInputClearButton",
  setup: function setup(_, _ref3) {
    var slots = _ref3.slots,
        attrs = _ref3.attrs;
    var api = usePinInput();
    return function () {
      return createVNode(chakra.button, mergeProps({
        "__label": "pin-input-clear-button",
        "onClick": function onClick() {
          return api.value.clearValue();
        }
      }, attrs), {
        "default": function _default() {
          return getValidChildren(slots);
        }
      });
    };
  }
});

export { CPinInput, CPinInputClearButton, CPinInputField, CPinInputProps, PinInputProvider, usePinInput };
