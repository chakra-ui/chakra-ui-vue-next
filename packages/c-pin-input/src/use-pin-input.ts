import { computed, watch } from "vue"
import { connect, machine } from "@zag-js/pin-input"
import { normalizeProps, useMachine } from "@zag-js/vue"

export interface PinInputProps {
  id: string
  value: string[]
  modelValue: string[]
  placeholder: string
  type: "alphanumeric" | "numeric"
  otp: boolean
  mask: boolean
  blurOnComplete: boolean
}

export const usePinInputMachine = (
  params: PinInputProps,
  emit: CallableFunction
) => {
  const [state, send] = useMachine(
    machine({
      id: params.id,
      value: params.modelValue || params.value,
      placeholder: params.placeholder,
      type: params.type,
      otp: params.otp,
      mask: params.mask,
      blurOnComplete: params.blurOnComplete,
      onChange: (value) => {
        emit("change", value)
        emit("update:modelValue", value)
      },
      onComplete: ({ value, valueAsString }) => {
        emit("complete", { value, valueAsString })
      },
      onInvalid: (details) => {
        emit("invalid", details)
      },
    }) as any
  )

  const api = computed(() => connect(state.value as any, send, normalizeProps))
  watch(
    () => params.modelValue,
    (value) => {
      api.value.setValue(Array.from(value))
    }
  )

  return api
}
