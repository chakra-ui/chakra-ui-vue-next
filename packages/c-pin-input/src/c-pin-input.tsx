/**
 * Hey! Welcome to @chakra-ui/vue-next CPinInput
 *
 * The pin input component is similar to the input component but is optimized for entering sequences of digits quickly
 *
 * @see Docs     https://next.vue.chakra-ui.com/c-pin-input
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue-next/blob/master/packages/c-pin-input/src/c-pin-input/c-pin-input.tsx
 * @see WAI-ARIA https://www.w3.org/TR/wai-aria-practices-1.2
 */

import { h, defineComponent, computed, PropType, ComputedRef } from "vue"
import {
  chakra,
  createStylesContext,
  useStyleConfig,
} from "@chakra-ui/vue-system"
import {
  createContext,
  getValidChildren,
  vueThemingProps,
} from "@chakra-ui/vue-utils"
import { CInput } from "@chakra-ui/c-input"
import { connect } from "@zag-js/pin-input"
import { PinInputProps, usePinInputMachine } from "./use-pin-input"

const [StylesProvider, useStyles] = createStylesContext("CPinInput")
const [PinInputProvider, usePinInput] =
  createContext<ComputedRef<ReturnType<typeof connect>>>()

export { PinInputProvider, usePinInput }

export const CPinInputProps = {
  value: {
    type: Object as PropType<Array<string>>,
    default: [],
  },
  modelValue: {
    type: Object as PropType<Array<string>>,
    default: false,
  },
  id: {
    type: String,
    default: "0",
  },
  placeholder: {
    type: String,
    default: "o",
  },
  type: {
    type: String as PropType<"alphanumeric" | "numeric">,
    default: "numeric",
  },
  otp: {
    type: Boolean,
    default: false,
  },
  mask: {
    type: Boolean,
    default: false,
  },
  blurOnComplete: {
    type: Boolean,
    default: false,
  },
  dir: {
    type: String as PropType<"rtl" | "ltr">,
    default: "ltr",
  },
  spacing: {
    type: [String, Number],
    default: "0.75",
  },
  ...vueThemingProps,
}

export const CPinInput = defineComponent({
  name: "CPinInput",
  props: CPinInputProps,
  emits: ["change", "invalid", "complete", "update:modelValue"],
  setup(props, { slots, attrs, emit }) {
    const styles = useStyleConfig("PinInput", props)

    const inputStyles = computed(() => ({
      ...styles.value,
      mx: props.spacing,
    }))
    StylesProvider(inputStyles)

    const api = usePinInputMachine(props as unknown as PinInputProps, emit)

    PinInputProvider(api)

    return () => (
      <chakra.div __label="pin-input" {...api.value.rootProps} {...attrs}>
        {() =>
          getValidChildren(slots).map((child, index) =>
            (child.type as any).name === "CPinInputField"
              ? h(child, { index })
              : child
          )
        }
      </chakra.div>
    )
  },
})

export const CPinInputField = defineComponent({
  name: "CPinInputField",
  props: {
    index: {
      type: Number,
      default: 0,
    },
  },
  setup(props, { attrs }) {
    const styles = useStyles()
    const api = usePinInput()

    return () => (
      <CInput
        __label="pin-input-field"
        __css={styles.value}
        {...api.value.getInputProps({ index: props.index })}
        {...attrs}
      />
    )
  },
})

export const CPinInputClearButton = defineComponent({
  name: "CPinInputClearButton",
  setup(_, { slots, attrs }) {
    const api = usePinInput()
    return () => (
      <chakra.button
        __label="pin-input-clear-button"
        onClick={() => api.value.clearValue()}
        {...attrs}
      >
        {() => getValidChildren(slots)}
      </chakra.button>
    )
  },
})
