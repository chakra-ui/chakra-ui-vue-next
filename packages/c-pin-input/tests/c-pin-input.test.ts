import { CPinInput, CPinInputField } from "./../src"
import { render } from "../../test-utils/src"

const renderCPinInput = (props?) => {
  const template = `
    <CPinInput>
      <CPinInputField />
      <CPinInputField/>
      <CPinInputField/>
    </CPinInput>`
  const base = {
    components: {
      CPinInput,
      CPinInputField,
    },
    template,
    ...props,
  }
  return render(base)
}

describe("CPinInput", () => {
  it("should render properly", () => {
    const { asFragment } = renderCPinInput()
    expect(asFragment()).toMatchSnapshot()
  })
})
