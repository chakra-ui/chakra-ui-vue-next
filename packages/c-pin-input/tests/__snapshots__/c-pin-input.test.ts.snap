// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CPinInput should render properly 1`] = `
<DocumentFragment>
  <div
    class="chakra-pin-input css-0"
    data-complete=""
    data-part="root"
    id="pin-input:0"
  >
    <input
      aria-label="pin code 1 of 0"
      autocapitalize="none"
      autocomplete="off"
      class="chakra-pin-input-field css-11ncskj"
      data-complete=""
      data-ownedby="pin-input:0"
      data-part="input"
      id="pin-input:0:0"
      inputmode="numeric"
      placeholder="o"
      type="tel"
    />
    <input
      aria-label="pin code 2 of 0"
      autocapitalize="none"
      autocomplete="off"
      class="chakra-pin-input-field css-11ncskj"
      data-complete=""
      data-ownedby="pin-input:0"
      data-part="input"
      id="pin-input:0:1"
      inputmode="numeric"
      placeholder="o"
      type="tel"
    />
    <input
      aria-label="pin code 3 of 0"
      autocapitalize="none"
      autocomplete="off"
      class="chakra-pin-input-field css-11ncskj"
      data-complete=""
      data-ownedby="pin-input:0"
      data-part="input"
      id="pin-input:0:2"
      inputmode="numeric"
      placeholder="o"
      type="tel"
    />
  </div>
</DocumentFragment>
`;
